src/System.Management.Automation/engine/hostifaces/PowerShell.cs

```diff
@@ -860,6 +860,20 @@ public static PowerShell Create(InitialSessionState initialSessionState)
             return result;
         }
 
+        /// <summary>
+        /// Constructs an empty PowerShell instance; a script or command must be added before invoking this instance.
+        /// </summary>
+        /// <param name="runspace">Runspace in which to invoke commands.</param>
+        /// <returns>An instance of PowerShell.</returns>
+        public static PowerShell Create(Runspace runspace)
```

Comment:
I think this will be very useful.  However, we should always validate parameter arguments on public methods.
```
if (runspace == null)
{
    throw new System.Management.Automation.PSArgumentNullException("runspace");
}
```
I agree that the provided runspace does not need to be in an "Opened" state, however, if it is opened then we should check its Availability.  It is useless if the Availability is "Busy" or "RemoteDebug".  If it is "AvailableForNestedCommand" then the IsNested property must be true, but you can run a nested command only if the PowerShell object is running on the same thread.  You can reuse existing error messages.

Also, if RunspaceState is "Disconnected", then this is useless ... there is a separate method for getting a PowerShell object that can reconnect to a remote session.  Similar issue for "Broken" or "Closed" states.  I think it would be good to provide the user an error immediately when the runspace is clearly not in a valid state for assignment to the PowerShell object.  "BeforeOpen" state is Ok because that can be corrected.  But the other states, there is nothing the use can do and I feel an error should be immediately thrown.

Reply:
Seems other create methods don't check runspace state.

We should use `nameof` in `PSArgumentNullException(nameof(runspace))`.

Reply:
"Seems other create methods don't check runspace state."

I guess it is a matter of opinion.  You can create a PowerShell object without any runspace association at all, but it seems to me that a static method that specifically constructs a PowerShell object for a provided runspace, should ensure that the returned object is viable.  I think of it as input parameter validation, if the runspace is not usable then it is not a valid input to the method.

Reply:
Code can't predict the future.  It can only work with the data on hand.  It seems wrong to me to successfully return an object known to be invalid.  The user then spends effort composing a possibly complex serious of statements only to have it inevitably fail.  As a user I prefer to know that the returned object is invalid up front.

But I am Ok with just checking for null exception on the runspace parameter argument.

Reply:
I see your point. My thoughts was that we create the runspace object _before_ Powershell object and can/should check a state of the runspace before creating Powershell object. I'd expect such user behavior.
 I am Ok with just checking for null exception too.

Reply:
Thanks for this great discussion guys. I really like the null check because the intent here is very clearly to use an existing runspace, so having it check against null makes great sense. I added a null check and a Pester test to ensure it throws when passed null.